### ISSUE TITLE
fragmented event emitting strategy

### ISSUE THREAD

**Issue by cdaringe** [pouchdb/pouchdb#4927]
#### problem statements
generally, the event structure to get DB activity is scattered among various internal APIs and not easily available to use at any point in a pouch instance lifecycle.

- db sync events are _exclusively available_ from the emitter returned when calling `sync`
    - events are _not_ available on a PouchDB instance, even if some script called `.sync` on a PouchDB instance elsewhere
- db sync events are _not_ available on the `.replicate` api 
- db change events are _exclusively available_ from the emitter returned from the `db.changes(...)` api
- created/destroyed events are always attached to the global PouchDB, vs. pouches which may be connected to various remote CouchDB instances.
    - this may be a non-issue, but it's peculiar that to get events, you must first have an instance.  Therefore, it would logical to assume that created/destroyed events could come from the instance, not the class (for lack of better terms)

#### solution
- make each PouchDB instance an EventEmitter
- enable sync/replicate to simply `this.emit` on instance emitter
- enable changes to simply `this.emit`  on instance emitter
- move sync events inside of replicate, s.t. both apis emit

#### discussion
this is just brainstorming, but I have needs where this would really be helpful.  i often have to return to the pouch docs to ask "how do I get that type of event again?" because events i'm interested in come from three different channels.  I think it would simplify the event structure and improve usability.  events can come from a single emitter, and are always easily accessible.  for example, my code is not the code calling `db.changes(...)` (maybe my co-workers owned code, etc), but i definitely want to subscribe to them.  i can of course, stash emitters somewhere in state, but it feels redundant when `db.on('some-event', () => ...)` could be so natural.

i know this would be a huge undertaking, code wise, doc wise, and test-wise.  regardless, perhaps a full major version bump cycle could be explicitly focused on just the one goal.  sorry for the long writeup!  it's been on my mind for a while!  i will contribute if the team is interested!

**Comment by nolanlawson** [comment]
Part of the reason we've avoided attaching events to the `PouchDB` object is because EventEmitters do have a memory overhead, and "max number of EventEmitters exceeded" errors have been historically very common (especially when we accidentally introduced leaks).

I understand your frustration with dealing with multiple EventEmitters, but I think if you consider some edge cases you'll see why it's designed that way:

* Multiple ongoing repilcations - which `'change'` events to emit?
* Remote databases - if users aren't going to listen to changes feed, no sense in opening a permanent HTTP connection
* Docs - user hasn't declared whether they want `include_docs:true` or not (this has a performance impact).
* Attachments - user hasn't declared whether or not they want attachment data to be emitted, and in what format (binary/base64). This has a performance impact.

So based on these cases, I would be inclined to say that the existing system should probably stay as-is. If the API is difficult to understand, I think maybe that would be an argument for cleaning up the documentation page, rather than potentially increasing the API surface or introducing breaking changes. But that's just my opinion; I'll let @daleharvey chime in too if he wants.

**Comment by cdaringe** [comment]
@nolanlawson, thanks for the thoughtful response!  my response is long, simply because i'm thinking out-loud as I digest your remarks!

> EventEmitters do have a memory overhead

yea, def.  i was thinking that it could be lazily instantiated.  e.g., when people want events, `myPouch._events` gets instantiated, and all other emitters route through there

> Multiple ongoing replications - which 'change' events to emit?

good point.  well, I dont think there'd need to be _too_ much change from the status quo. instead of getting events exclusively from the `changes` API, a call to `changes()` simply `UPSERT`s `myPouch._events`, and proxies events through there. if you `replicate.to(x)` and `replicate.to(y)`... yea, the event payload  would need some additional data: `{ event: "change", db: "x", doc: {...} }`.

what i suggest ^^ is something like:

```
myPouch._events (Emitter)
^ pipes events      ^ pipes events    ^ pipes events             ^ pipes events
dbA-syncEmitter  dbB-syncEmitter db-created/destroyed  doc-changes

==>
myPouch.changes(...);
myPouch.on('change', () => { ... });
myPouch.sync(...).on('sync', (evt) => { if (evt.complete/inProgress) { ... } }); // where .sync _returns_ myPouch._events, and you can chain right on!
```

> Remote databases - if users aren't going to listen to changes feed, no sense in opening a permanent HTTP connection

absolutely, completely agree.  again with the `lazy` bit!

> Docs - ... Attachments ...

could be a no-op if the proxy solution was considered

----

to summarize again, events happen on my stores.  sync events, change events, etc.  i must cache Pouch generated emitters in my app and pass them around to those who need to listen.  a more natural feeling approach to is enable `myPouch.on(...)` like interface.  this provides simple and unified Event API for a pouch instance.

**Comment by nolanlawson** [comment]
Thanks for the explanation, although I admit I'm still wary to add these features to PouchDB core. IMO this problem can be solved in plugin-land, or as a wrapper around PouchDB, with zero performance penalty and no need to make breaking changes to the API.

E.g.

```js
var EventfulPouch = require('eventful-pouch');
var db = new EventfulPouch('name');
db.on(...);
```

^ and that object could automatically do all the event handling itself and even lazily, if it so chose. Events can be proxied, e.g. as we do in [`sync.js`](https://github.com/pouchdb/pouchdb/blob/9344d555b37b63bd7e0664b928e6cee9cdb40530/src/sync.js).

**Comment by nolanlawson** [comment]
Closing because, again, I think it's super important that PouchDB core be kept small and fairly consistent. The API has reached a point of (relative) stability now, with a plugin API that allows most cosmetic changes to be handled in userland (e.g. see [pouchy](https://github.com/cdaringe/pouchy)). I definitely see your idea as something that can grow the ecosystem rather than stretch the responsibilities of the core API.

I strongly encourage you to write the module with the API as you describe. :) But I think keeping Pouch core as-is (despite the API not being perfectly intuitive) is the best approach.
